File: app/src/main/java/me/instabattle/app/models/Battle.java
package me.instabattle.app.models;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import com.google.android.gms.maps.model.LatLng;

import java.util.Date;
import java.util.List;

import me.instabattle.app.managers.EntryManager;
import me.instabattle.app.managers.UserManager;
import me.instabattle.app.settings.State;
import retrofit2.Callback;

public class Battle {
    @SerializedName("radius")
    @Expose
    private Integer radius;
    @SerializedName("entry_count")
    @Expose
    private Integer entriesCount = 0;
    @SerializedName("created_on")
    @Expose
    private Date createdOn;
    @SerializedName("description")
    @Expose
    private String description;
    @SerializedName("id")
    @Expose
    private Integer id;
    @SerializedName("latitude")
    @Expose
    private Double latitude;
    @SerializedName("longitude")
    @Expose
    private Double longitude;
    @SerializedName("name")
    @Expose
    private String name;
    @SerializedName("creator_id")
    @Expose
    private Integer creatorId;


    public Battle(Integer creatorId, String name, String description,
                  Double lat, Double lng, Integer radius) {
        this.creatorId = creatorId;
        this.name = name;
        this.description = description;
        this.latitude = lat;
        this.longitude = lng;
        this.radius = radius;

        this.creatorId = State.currentUser.getId();
        this.entriesCount = 1;
        this.createdOn = new Date();
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public Date getCreatedOn() {
        return createdOn;
    }

    public LatLng getLocation() {
        return new LatLng(latitude, longitude);
    }

    public int getRadius() {
        return radius;
    }

    public void getEntriesAndDo(Callback<List<Entry>> callback) {
        EntryManager.getByBattleAndDo(id, callback);
    }

    public int getEntriesCount() {
        return entriesCount;
    }

    public void getWinnerAndDo(Callback<List<Entry>> callback) {
        EntryManager.getTopByBattleAndDo(id, 1, callback);
    }

    public void getVoteAndDo(Callback<List<Entry>> callback) {
        EntryManager.getVoteAndDo(id, callback);
    }

    public void createEntryAndDo(Callback<Entry> callback) {
        EntryManager.createAndDo(id, State.currentUser.getId(), callback);
    }

    public void getAuthorAndDo(Callback<User> callback) {
        UserManager.getAndDo(creatorId, callback);
    }

    public Integer getId() {
        return id;
    }
}

File: app/src/main/kotlin/me/instabattle/app/models/Battle.kt
package me.instabattle.app.models

import android.os.Parcelable
import com.google.android.gms.maps.model.LatLng
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import kotlinx.android.parcel.Parcelize
import me.instabattle.app.managers.EntryManager
import me.instabattle.app.managers.UserManager
import me.instabattle.app.settings.GlobalState
import retrofit2.Callback
import java.util.*

@Parcelize
data class Battle(
        @field:Expose @field:SerializedName("user_id")
        private val userId: Int,
        @field:Expose @field:SerializedName("name")
        val name: String,
        @field:Expose @field:SerializedName("description")
        val description: String,
        @field:Expose @field:SerializedName("latitude")
        private val latitude: Double?,
        @field:Expose @field:SerializedName("longitude")
        private val longitude: Double?,
        @field:Expose  @field:SerializedName("radius")
        val radius: Int?,
        @field:Expose @field:SerializedName("entry_count")
        var entriesCount: Int = 0,
        @field:Expose @field:SerializedName("created_on")
        var createdOn: Date? = null,
        @field:Expose @field:SerializedName("id")
        val id: Int? = null,
        @field:Expose @field:SerializedName("creator_id")
        private var creatorId: Int? = null ): Parcelable {

    val location: LatLng
        get() = LatLng(latitude!!, longitude!!)

    init {
        this.creatorId = userId

        this.creatorId = GlobalState.currentUser.id
        this.entriesCount = 1
        this.createdOn = Date()
    }//FIXME

    fun getEntriesAndDo(callback: Callback<List<Entry>>) = EntryManager.getByBattleAndDo(id!!, callback)

    fun getWinnerAndDo(callback: Callback<List<Entry>>) = EntryManager.getTopByBattleAndDo(id!!, 1, callback)

    fun getVoteAndDo(callback: Callback<List<Entry>>) = EntryManager.getVoteAndDo(id, callback)

    fun createEntryAndDo(callback: Callback<Entry>) = EntryManager.createAndDo(id, GlobalState.currentUser.id, callback)

    fun getAuthorAndDo(callback: Callback<User>) = UserManager.getAndDo(creatorId, callback)
}
